src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.Rewriter.cs

```diff
@@ -244,14 +244,25 @@ private ExpressionSyntax RewriteSwitchStatement(SwitchStatementSyntax node, bool
                     }
                 }
 
+                var switchExpression = node.Expression;
+                if (MustParentheseExpression(switchExpression))
```

Comment:
I believe a better approach would be to _always_ parenthesize using [ISyntaxFactsService.Parenthesize](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs,0c90028a7d30e5a3,references) API, which adds a simplifier annotation by default, and then let the simplifier post pass take care of removing unnecessary parentheses based on user's parentheses preference.

```csharp
return SwitchExpression(
    syntaxFacts.Parenthesize(node.Expression),
    ...
```

Reply:
I tried to use the extension method `ExpressionSyntaxExtensions.Parenthesize(...)` (which looks equivalent to `ISyntaxFactsService` as the rewriter is C# specific) however 2 tests are failing.

It fails when the rewritten expression looks like

```csharp
var j = (i) switch { ... }
```

Roslyn tries to bind a deconstruction and fails on an assert:

````csharp
            DeconstructionVariable locals = BindDeconstructionVariables(left, diagnostics, ref declaration, ref expression);
            Debug.Assert(locals.HasNestedVariables);
````

````
    System.InvalidOperationException : 
    
  Stack Trace: 
    at ThrowingTraceListener.Fail(String message, String detailMessage) in ThrowingTraceListener.cs line: 24
    at TraceListener.Fail(String message)
    at TraceInternal.Fail(String message)
    at Debug.Assert(Boolean condition)
    at Binder.BindDeconstruction(CSharpSyntaxNode deconstruction, ExpressionSyntax left, ExpressionSyntax right, DiagnosticBag diagnostics, DeclarationExpressionSyntax& declaration, ExpressionSyntax& expression, Boolean resultIsUsedOverride, BoundDeconstructValuePlaceholder rightPlaceholder) in Binder_Deconstruct.cs line: 95
    at Binder.BindDeconstruction(AssignmentExpressionSyntax node, DiagnosticBag diagnostics, Boolean resultIsUsedOverride) in Binder_Deconstruct.cs line: 33
    at Binder.BindAssignment(AssignmentExpressionSyntax node, DiagnosticBag diagnostics) in Binder_Statements.cs line: 1333
    at Binder.BindExpressionInternal(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed) in Binder_Expressions.cs line: 404
    at Binder.BindExpression(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed) in Binder_Expressions.cs line: 329
    at Binder.BindValue(ExpressionSyntax node, DiagnosticBag diagnostics, BindValueKind valueKind) in Binder_Expressions.cs line: 228
    at Binder.BindExpressionStatement(CSharpSyntaxNode node, ExpressionSyntax syntax, Boolean allowsAnyExpression, DiagnosticBag diagnostics) in Binder_Statements.cs line: 588
    at Binder.BindExpressionStatement(ExpressionStatementSyntax node, DiagnosticBag diagnostics) in Binder_Statements.cs line: 581
    at Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics) in Binder_Statements.cs line: 52
    at Binder.BindBlockParts(BlockSyntax node, DiagnosticBag diagnostics) in Binder_Statements.cs line: 1662
    at Binder.BindBlock(BlockSyntax node, DiagnosticBag diagnostics) in Binder_Statements.cs line: 1650
    at Binder.BindStatement(StatementSyntax node, DiagnosticBag diagnostics) in Binder_Statements.cs line: 43
    at Binder.BindMethodBody(CSharpSyntaxNode declaration, BlockSyntax blockBody, ArrowExpressionClauseSyntax expressionBody, DiagnosticBag diagnostics) in Binder_Statements.cs line: 3170
    at Binder.BindMethodBody(CSharpSyntaxNode syntax, DiagnosticBag diagnostics) in Binder_Statements.cs line: 3118
    at MethodCompiler.BindMethodBody(MethodSymbol method, TypeCompilationState compilationState, DiagnosticBag diagnostics, ImportChain& importChain, Boolean& originalBodyNested, ValueTuple`4& forSemanticModel) in MethodCompiler.cs line: 1653
    at MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState) in MethodCompiler.cs line: 958
    at MethodCompiler.CompileNamedType(NamedTypeSymbol containingType) in MethodCompiler.cs line: 507
    at MethodCompiler.VisitNamedType(NamedTypeSymbol symbol, TypeCompilationState arg) in MethodCompiler.cs line: 385
    at NamedTypeSymbol.Accept[TArgument,TResult](CSharpSymbolVisitor`2 visitor, TArgument argument) in NamedTypeSymbol.cs line: 576
    at MethodCompiler.CompileNamespace(NamespaceSymbol symbol) in MethodCompiler.cs line: 364
    at MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean emittingPdb, Boolean emitTestCoverageData, Boolean hasDeclarationErrors, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in MethodCompiler.cs line: 152
    at CSharpCompilation.GetDiagnosticsForMethodBodiesInTree(SyntaxTree tree, Nullable`1 span, CancellationToken cancellationToken) in CSharpCompilation.cs line: 2412
    at CSharpCompilation.GetDiagnosticsForSyntaxTree(CompilationStage stage, SyntaxTree syntaxTree, Nullable`1 filterSpanWithinTree, Boolean includeEarlierStages, CancellationToken cancellationToken) in CSharpCompilation.cs line: 2544
    at SyntaxTreeSemanticModel.GetDiagnostics(Nullable`1 span, CancellationToken cancellationToken) in SyntaxTreeSemanticModel.cs line: 134
    at <GetErrorsAsync>d__7.MoveNext() in DocumentExtensions.cs line: 120
...
````

Reply:
That's a separate issue (a bug in the compiler).  We should definitely fix that bug.  However, we should also just call the `.Parenthesize()` extension method which already exists (and which will annotate properly to have the parentheses removed if not necessary.

Reply:
Tagging @gafter @jcouv about the above compiler assert.

Reply:
@meziantou Just following up to check if you were able to file a new issue for the assert seen above.

Reply:
Seems like there are 2 tests failing currently with this assert. @meziantou Would you be able to file an issue and skip the failing tests with this issue? I can merge the PR once the checks are green.

Reply:
I'll do that tomorrow!

Reply:
Here's the new issue: https://github.com/dotnet/roslyn/issues/37444. I've also skipped the 2 tests.

Reply:
There is a new failed test in the VS integration tests. I'm not sure if it is related to my changes.

````
Roslyn.VisualStudio.IntegrationTests.CSharp.CSharpInteractiveCommands.VerifyPreviousAndNextHistory [FAIL]
````

Reply:
It looks like you've constructed an ill-formed syntax tree. You've constructed an expression statement containing a declaration expression on the left. That is not legal and the compiler is right to crash on a tree that the parser could never produce. This should be a local declaration statement with initializer.

Reply:
Thanks @gafter for the analysis. I've changed the way the variable is declared and the tests succeed!